library/src/animationManager.cxx

```diff
@@ -212,18 +205,14 @@ void animationManager::Tick()
   // Convert to a usable time in seconds
   double elapsedTime = static_cast<double>(timeInMS) / 1000.0;
   double animationSpeedFactor = this->Options->getAsDouble("scene.animation.speed-factor");
-  elapsedTime *= animationSpeedFactor;
 
   // elapsedTime can be negative
-  this->CurrentTime += elapsedTime;
-  if (this->CurrentTime < this->TimeRange[0])
-  {
-    this->CurrentTime = this->TimeRange[1];
-  }
-  else if (this->CurrentTime > this->TimeRange[1])
-  {
-    this->CurrentTime = this->TimeRange[0];
-  }
+  elapsedTime *= animationSpeedFactor;
+
+  // Modulo computation, compute CurrentTime + elapsedTime in the time range.
+  double delta = this->TimeRange[1] - this->TimeRange[0];
+  this->CurrentTime = fmod(this->CurrentTime - this->TimeRange[0] + fmod(elapsedTime, delta) + delta, delta) + this->TimeRange[0];
```

Comment:
```suggestion
  this->CurrentTime = fmod(fmod(this->CurrentTime - this->TimeRange[0], delta) + delta, delta) + this->TimeRange[0];
```

We could really do with a proper modulo function defined somewhere (it came up before):
```cpp
double modulo(double a, double b)
{
    const double remainder = fmod(a, b);
    return remainder < 0 ? remainder + b : remainder;
}
```
because the whole `fmod(fmod(a, b)+b, b)` thing is clever but ugly, then:
```suggestion
  this->CurrentTime = this->TimeRange[0] + modulo(this->CurrentTime - this->TimeRange[0], this->TimeRange[1] -  this->TimeRange[0]);
```

Also I think that may not work when exactly at the bounds so it'd need an `if(this->CurrentTime < this->TimeRange[0] || this->CurrentTime > this->TimeRange[1])` first

Reply:
good point, especially I hurt my brain to write it. Ill take care of that.

Reply:
> (it came up before)

Do you remember where ? Cant find it.

I will add it locally for now.

Reply:
https://github.com/f3d-app/f3d/blob/master/application/F3DStarter.cxx#L412-L413

Reply:
> Also I think that may not work when exactly at the bounds

It behave as expected on this->TimeRange[0]

On this->TimeRange[1], it wraps to this->TimeRange[0], which I guess we may not want.

Reply:
> > Also I think that may not work when exactly at the bounds
> 
> It behave as expected on this->TimeRange[0]
> 
> On this->TimeRange[1], it wraps to this->TimeRange[0], which I guess we may not want.

ah yes, because `fmod(x,x) == 0` so the lower bound would be included, but the upper bound would be excluded.

So technically only `if(this->CurrentTime > this->TimeRange[1])` is necessary, but `if(this->CurrentTime < this->TimeRange[0] || this->CurrentTime > this->TimeRange[1])` wouldn't hurt either, up to you

Reply:
that's what I did :)

Reply:
> https://github.com/f3d-app/f3d/blob/master/application/F3DStarter.cxx#L412-L413

This one is int based and way less complex, not sure that's worth adding a (private) shared namespace for that yet ?